servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpPayloadDiscardWatchdogServiceFilter.java

```diff
@@ -0,0 +1,84 @@
+import io.servicetalk.context.api.ContextMap;
+import io.servicetalk.http.api.HttpApiConversions;
+import io.servicetalk.http.api.HttpExecutionStrategies;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+/**
+ * Filter which tracks HTTP payloads sent by the service so it can be freed if discarded in the pipeline.
+ */
+final class HttpPayloadDiscardWatchdogServiceFilter implements StreamingHttpServiceFilterFactory {
+
+    /**
+     * Instance of {@link HttpPayloadDiscardWatchdogServiceFilter}.
+     */
+    static final StreamingHttpServiceFilterFactory INSTANCE = new HttpPayloadDiscardWatchdogServiceFilter();
+
+    @SuppressWarnings("rawtypes")
+    static final ContextMap.Key<Publisher> payloadPublisherKey = ContextMap.Key
+            .newKey("io.servicetalk.http.netty.payloadPublisher", Publisher.class);
+
+    static final ContextMap.Key<Boolean> payloadSubscribedKey = ContextMap.Key
+            .newKey("io.servicetalk.http.netty.payloadSubscribed", Boolean.class);
+
+    private HttpPayloadDiscardWatchdogServiceFilter() {
+        // Singleton
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(final StreamingHttpService service) {
+
+        return new StreamingHttpServiceFilter(service) {
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        final StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                return delegate()
+                        .handle(ctx, request, responseFactory)
+                        .map(response -> {
+                            if (!HttpApiConversions.isPayloadEmpty(response)) {
+                                // If the payload is not empty, always write the buffer publisher into the
+                                // request context. When a downstream subscriber arrives, mark the payload
+                                // as subscribed explicitly (having a payload present and no subscription is
+                                // an indicator that it must be freed later on).
+                                request.context().put(payloadPublisherKey, response.payloadBody());
+                                return response.transformPayloadBody(bufferPublisher ->
+                                        bufferPublisher.beforeOnSubscribe(subscription ->
+                                                request.context().put(payloadSubscribedKey, true)));
```

Comment:
Alternatively, consider using `AtomitReference` for the publisher and set it back to `null` on subscribe. That way we need only a single key for context map.
This will also help with retry case mentioned by Scott. If we see `response.map` execution again, while reference is not null, it's an indicator that we have to drain the previous publisher too.

Reply:
Added a test